### PR TITLE
temp PR for review

### PR DESCRIPTION
**Pull request by poltak**
@mukeshkharita

### CODE REVIEW

File: src/components/storage/aws.ts [hunk 1]
```diff
@@ -24,41 +20,26 @@ export class AwsUserStorage implements UserStorage {
...
-        let body;
-
-        await new Promise((resolve, reject) => {
-            stringify(users, function(err, output){
-                body = output;
-                resolve(output)
-            });
-        })
+    async storeUser(id:string, installTime: number) {
+        const key = 'users/' + id + '.json'
```
Comment (poltak) [inline src/components/storage/aws.ts:24]: There's lots of these string and object literals around this class. They should be set as properties on the class instance, so you could go `const key = this.dir + id + this.ext` (names are just a suggestion). 
Think about if you have to change any of these and then you need to search through your code to find all occurrences. Generally you would want them to be able to be set from the `constructor`, so external modules can set these. But you may not want to specify them every time you create an instance, so you could set their default values to be static class properties in the constructor args. 

tl;dr: use class props to DRY out all these literal values

Quick example (just a suggestion; add things as appropriate):

```ts
class Test {
  static DEF_DIR = 'users/'
  static DEF_EXT = '.json'
  static DEF_CONTENT_TYPE = {json: 'text/json', ...}

  constructor({bucketName, dir = Test.DEF_DIR, ext = Test.DEF_EXT, contentType = Test.DEF_CONTENT_TYPE}) {
    this.dir = dir
    this.ext = ext
    ... // init everything else
  }
}
```

On the topic of class props:
Instance props `_s3` and `bucketName` are both `public` but I don't see them being accessed outside the class methods. I would opt for making your instance props `private` by default unless you have a good reason not to (you can also make use of setters/getters for outside access. Good info: https://www.typescriptlang.org/docs/handbook/classes.html#accessors).

File: src/components/storage/aws.ts [hunk 2]
```diff
@@ -24,41 +20,26 @@ export class AwsUserStorage implements UserStorage {
...
-        })
-        
-        return true;
+    async userExists(id) {
+        const key = 'users/' + id + '.json'
+        return await this._getObject({key, type: 'json'})        
     }
 
-    async _putObject({key, body, type, mime} : {key : string, body, type? : 'csv', mime? : string}) {
+    async _putObject({key, body, type, mime} : {key : string, body: any, type? : 'json', mime? : string}) {
```
Comment (poltak) [inline src/components/storage/aws.ts:36]: This type signature is repeated in both the user and event storage classes. Same issue with the `_getObject` methods. Immediate improvement could be to DRY it out by doing something like:

```ts
interface PutObjectProps {
  key: string
  body: any,
  type?: 'json'   // Please note that this is saying `type` field can only ever be set to the value 'json' or undefined!
  mime?: string
}
```

and changing your method to something like

```ts
async _putObject({key, body, type, mime}: PutObjectProps) {...}
```

Even better would be creating an abstract class that both these classes extend. The abstract class would have abstract `_putObject` and `_getObject` methods, which then these classes would be forced to implement. More info on abstract classes: https://www.typescriptlang.org/docs/handbook/classes.html#abstract-classes

File: src/components/storage/aws.ts [hunk 3]
```diff
@@ -114,27 +91,30 @@ export class AwsEventLogStorage implements EventLogStorage {
...
+        for(let event of events.data) {
+            const body = {
+                time: event.time,
+                id: events.id,
+                other: event.other,
+                type: event.type
+            }
 
-        let body;
+            const eventId = event.time + event.type
```
Comment (poltak) [inline src/components/storage/aws.ts:104]: We talked about this briefly, but how big is the chance that an event of the same type happens on the same millisecond? What about adding in a string version of the `event.other` as part of the event ID? TBH I'm not really sure what this though (adding typings for events would be great here!).

File: src/components/storage/aws.ts [hunk 4]
```diff
@@ -148,35 +128,13 @@ export class AwsEventLogStorage implements EventLogStorage {
...
-            })
-        })
-    }
 
-    async _getObject({key, type} : {key : string, type? : 'csv'}) {
-        const params = {
-            Bucket: this.bucketName,
-            Key: key,
+        try {
+            const headCode = this._s3.putObject(params).promise()
```
Comment (poltak) [inline src/components/storage/aws.ts:133]: From what I understand `headCode` is going to be set to type `Promise` then this method will always immediately return `true` regardless of the outcome of the Promise. The `catch` scope will never be entered as you would need an `await` to wait until the Promise resolves or rejects. 
Same issue with both the classes, although the `_getObject` methods seem to wait for the Promise's outcome ok (although I would avoid assigning `headCode` variable then never using it - you could set up tslint to warn you about unused assignments).

In general I would avoid having these get/putObject methods return booleans depending on the outcome of the Promises. They already return a Promise (as they're async methods), so you could just replace your `return false` calls with `throw someError` (translates to `Promise.reject()` outside of asyncawait) then the callers can simply use `try...catch`/`.then().catch()` and decide on how to act (rather than waiting on the Promise to resolve to either `true` or `false` and the possibility of handling a rejection that is semantically different to `false`).

File: src/components/storage/aws.ts [hunk 5]
```diff
@@ -114,27 +91,30 @@ export class AwsEventLogStorage implements EventLogStorage {
...
-        const key = 'events/events.csv'
-
-        const allEvent = await this._getObject({key, type: 'csv'})
-
-        events.data.forEach((event) => {
-            allEvent.push([event.time, events.id, event.other, event.type])
-        })
+        let storeEventSuccess = true
+
+        for(let event of events.data) {
```
Comment (poltak) [inline src/components/storage/aws.ts:96]: `event` is never reassigned within the `for...of` scope, so it should be a `const` rather than `let`. Very minor issue, but something you can set up your linter to warn you

File: src/controllers/generate-token.ts [hunk 6]
```diff
@@ -3,18 +3,18 @@ import { TokenGenerator } from '../components/token-generator'
...
     return async function handleGenerateTokenRequest({installTime}) {
         let id = tokenGenerator.generateToken()
-
-        while(!(await AwsUserStorage.isUserById(id))) {
+        while((await userStorage.userExists(id))) {
             id = tokenGenerator.generateToken()
         }
 
-        await AwsUserStorage.storeUser(id, installTime)
+        const user = await userStorage.storeUser(id, installTime)
```
Comment (poltak) [inline src/controllers/generate-token.ts:15]: Rejections are never handled from this Promise, so if this endpoint is called and there is a problem storing the new user and `storeUser` rejects, I'm not sure if a response would ever be sent (or maybe there is a global error handler somewhere that handles it? This is a valid way of handling misc. errors in express). 
Same issue with `eventLog` and `uninstall` endpoints

File: tslint.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
+{
+    "extends": ["tslint-react", "tslint-config-prettier"],
```
Comment (poltak) [inline tslint.json:2]: You haven't installed `prettier`, so you shouldn't use this plugin. IIRC this plugin just auto-disables all rules that prettier will handle for you (so you won't get annoyed by the linter or get conflicts between linter and prettier), so you might be missing quite a lot of style lint warnings if you have this here.

**Review by poltak — COMMENTED** [review]
@mukeshkharita I've left some feedback on here. Most important ones are just code improvements you could make to the storage classes + some other little things I noticed.

One other small thing to note is that when I `npm install` a fresh copy of the repo, `package-lock.json` changes. Maybe you forgot to update it when you added some new packages to `package.json`? npm version 6.1.0

File: src/components/storage/aws.ts (hunk 2 above)
Comment (mukeshkharita) [reply src/components/storage/aws.ts:36]: Hey Jon, I think we can just make one function for both the classes (out of both classes), and we can use in the both of the classes. By this way the code will not repeated, otherwise in the abstract class, we will implement same function for both classes, make sense? :slightly_smiling_face:

File: src/components/storage/aws.ts (hunk 3 above)
Comment (mukeshkharita) [reply src/components/storage/aws.ts:104]: The `event.type` is reserved and this is JSON, so I don't think so we have to use this to make the id. BTW the same event cannot occur at the same time for the same user.

File: src/components/storage/aws.ts (hunk 2 above)
Comment (poltak) [reply src/components/storage/aws.ts:36]: I totally overlooked this, but yes, they're almost the same implementation! You could put the shared logic into a method of a new class which both the event and user AWS storage classes extend. You should probably make that method `protected` (currently it's `public` as that is TS default for methods if you don't specify a privacy modifier).

File: src/controllers/generate-token.ts (hunk 6 above)
Comment (mukeshkharita) [reply src/controllers/generate-token.ts:15]: We can use `try {}catch(){}` here, right?

Comment (poltak) [reply src/controllers/generate-token.ts:15]: Yes, or look into error handling in express. I know from past experience there is smart ways of handling general errors on all routes (global error handling), which may be appropriate here.
